core/src/main/java/org/apache/calcite/rel/rules/IntersectToSemiJoinRule.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.rules;
+
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Intersect;
+import org.apache.calcite.rel.core.JoinRelType;
+import org.apache.calcite.rel.logical.LogicalIntersect;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexUtil;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.tools.RelBuilder;
+
+import org.immutables.value.Value;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Planner rule that translates a {@link org.apache.calcite.rel.core.Intersect}
+ * to a series of {@link org.apache.calcite.rel.core.Join} that type is
+ * {@link org.apache.calcite.rel.core.JoinRelType#SEMI}.
+ */
+@Value.Enclosing
+public class IntersectToSemiJoinRule
+    extends RelRule<IntersectToSemiJoinRule.Config>
+    implements TransformationRule {
+
+  /** Creates an IntersectToSemiJoinRule. */
+  protected IntersectToSemiJoinRule(Config config) {
+    super(config);
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    final Intersect intersect = call.rel(0);
+    if (intersect.all) {
+      return; // nothing we can do
+    }
+
+    final RelBuilder builder = call.builder();
+    final RexBuilder rexBuilder = builder.getRexBuilder();
+    final RelDataType oriRowType = intersect.getRowType();
+
+    List<RelNode> inputs = intersect.getInputs();
+    if (inputs.size() != 2) {
+      return;
+    }
+
+    RelNode left = inputs.get(0);
+    RelNode right = inputs.get(1);
+
+    List<RexNode> conditions = new ArrayList<>();
+    int fieldCount = left.getRowType().getFieldCount();
+
+    // Converting the join condition to
+    // AND(=(COALESCE($0, 0), COALESCE($1, 0)), =(IS NULL($0), IS NULL($0))
```

Comment:
Use this convertion, we can use HashJoin

Reply:
isn't IS NOT DISTINCT FROM the same thing but simpler?

Reply:
> isn't IS NOT DISTINCT FROM the same thing but simpler?

IS NOT DISTINCT FROM is convert to a condition with OR, it won't convert to HashJoin, and result of computing  IS NOT DISTINCT FROM is wrong currently.  We can see https://issues.apache.org/jira/projects/CALCITE/issues/CALCITE-6904

Reply:
> isn't IS NOT DISTINCT FROM the same thing but simpler?

AND(=(COALESCE($0, 0), COALESCE($1, 0)), =(IS NULL($0), IS NULL($0)) also can be samplier to =($0,$1) when both of them IS NOT NULL now.

Reply:
I don't understand what you are saying.
IS NOT DISTINCT FROM is really equality testing, where NULL is equal only to NULL.
What does this have anything to do with OR?
This stage of the optimizer is completely agnostic to the backend that will be used to generate code, so the issue about enumerable is not relevant.

Reply:
> I don't understand what you are saying. IS NOT DISTINCT FROM is really equality testing, where NULL is equal only to NULL. What does this have anything to do with OR? This stage of the optimizer is completely agnostic to the backend that will be used to generate code, so the issue about enumerable is not relevant.

I mean IS NOT DISTINCT FROM will be samplier to x=y is TRUE OR ((x is null) and (y is null)), this condition won't be used HashJoin.

Reply:
I think IS NOT DISTINCT FROM is exactly what Java calls `==`.

Reply:
Another problem is maybe a bug as I said in jira 6904.We can't get right result with IS NOT DISTINCT FROM currently. EnumableHashJoinRule make wrong convertion, the `filterNulls` is ignored.

Reply:
you do not make this rule more complex because another rule is wrong.
you should fix the rule which is wrong.

Reply:
> I think IS NOT DISTINCT FROM is exactly what Java calls `==`.

Yes, I agree with you! But the jdbcTest can't pass. We may remove the test temporarily.

Reply:
> you do not make this rule more complex because another rule is wrong. you should fix the rule which is wrong.

Thanks, I will change to IS NOT DISTINCT FROM, and remove JdbcTest.

Reply:
the first will work if I am right, you should double check.
removing tests is counterproductive, but you can temporarily disable a test.
there are many examples - see the Bug class.
But ideally the fix for the bug should be merged first...

Reply:
Got it. But for jira 6904 I can't find a best way to fix it.